### PR TITLE
add ssh support

### PR DESCRIPTION
**Pull request by ericsciple**
related to ADR https://github.com/actions/checkout/pull/156

### CODE REVIEW

File: src/git-source-settings.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+export interface IGitSourceSettings {
```
Comment (ericsciple) [inline src/git-source-settings.ts]: i moved this interface into its own file to reduce the amount of code in git-source-provider.ts

File: src/state-helper.ts [hunk 2]
```diff
@@ -12,6 +11,17 @@ export const IsPost = !!process.env['STATE_isPost']
 export const RepositoryPath =
   (process.env['STATE_repositoryPath'] as string) || ''
 
+/**
```
Comment (ericsciple) [inline src/state-helper.ts]: used during POST to cleanup the files on disk

File: src/git-source-provider.ts [hunk 3]
```diff
@@ -172,138 +163,6 @@ async function getGitCommandManager(
     }
 
     // Otherwise fallback to REST API
-    return (null as unknown) as IGitCommandManager
-  }
-}
-
-async function prepareExistingDirectory(
```
Comment (ericsciple) [inline src/git-source-provider.ts]: prepareExistingDirectory moved to a separate file

File: src/git-source-provider.ts [hunk 4]
```diff
@@ -172,138 +163,6 @@ async function getGitCommandManager(
...
-    // Delete the contents of the directory. Don't delete the directory itself
-    // since it might be the current working directory.
-    core.info(`Deleting the contents of '${repositoryPath}'`)
-    for (const file of await fs.promises.readdir(repositoryPath)) {
-      await io.rmRF(path.join(repositoryPath, file))
-    }
-  }
-}
-
-async function configureAuthToken(
```
Comment (ericsciple) [inline src/git-source-provider.ts]: the auth functions moved to a separate file

File: src/git-source-provider.ts [hunk 5]
```diff
@@ -92,12 +84,10 @@ export async function getSource(settings: ISourceSettings): Promise<void> {
...
 
-    // Remove possible previous extraheader
-    await removeGitConfig(git, authConfigKey)
-
+    const authHelper = gitAuthHelper.createAuthHelper(git, settings)
     try {
-      // Config extraheader
-      await configureAuthToken(git, settings.authToken)
+      // Configure auth
+      await authHelper.configureAuth()
```
Comment (ericsciple) [inline src/git-source-provider.ts]: the auth helper `configureAuth` also handles removing possible previous extraheader

File: src/git-directory-helper.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
+import * as core from '@actions/core'
```
Comment (ericsciple) [inline src/git-directory-helper.ts]: todo add unit tests for this file

File: __test__/git-auth-helper.test.ts [hunk 7]
```diff
@@ -0,0 +1,488 @@
+import * as core from '@actions/core'
```
Comment (ericsciple) [inline __test__/git-auth-helper.test.ts]: lots of tests for the auth helper

File: src/git-auth-helper.ts [hunk 8]
```diff
@@ -0,0 +1,201 @@
+import * as assert from 'assert'
```
Comment (ericsciple) [inline src/git-auth-helper.ts]: moved auth helper into a separate file so it can be thoroughly unit tested

File: src/git-auth-helper.ts [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+    let sshCommand = `"${sshPath}" -i "$RUNNER_TEMP/${path.basename(
+      this.sshKeyPath
+    )}"`
+    if (this.settings.sshStrict) {
+      sshCommand += ' -o StrictHostKeyChecking=yes -o CheckHostIP=no'
+    }
+    sshCommand += ` -o "UserKnownHostsFile=$RUNNER_TEMP/${path.basename(
+      this.sshKnownHostsPath
+    )}"`
+    this.git.setEnvironmentVariable('GIT_SSH_COMMAND', sshCommand)
```
Comment (ericsciple) [inline src/git-auth-helper.ts]: GIT_SSH_COMMAND takes precedence over git config `core.sshCommand`

File: src/git-directory-helper.ts (hunk 6 above)
Comment (ericsciple) [reply src/git-directory-helper.ts]: i have a bunch written, writing a few more still to increase coverage

File: src/git-auth-helper.ts [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+    // Configure core.sshCommand
+    if (this.settings.persistCredentials) {
+      await this.git.config(SSH_COMMAND_KEY, sshCommand)
```
Comment (ericsciple) [inline src/git-auth-helper.ts]: this runs `git config --local core.sshCommand "ssh -i path-to-key -o etc"`

File: src/git-auth-helper.ts [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+    }
+  }
+
+  private async configureToken(): Promise<void> {
+    // Configure a placeholder value. This approach avoids the credential being captured
+    // by process creation audit events, which are commonly logged. For more information,
+    // refer to https://docs.microsoft.com/en-us/windows-server/identity/ad-ds/manage/component-updates/command-line-process-auditing
+    const placeholder = `AUTHORIZATION: basic ***`
+    await this.git.config(EXTRA_HEADER_KEY, placeholder)
```
Comment (ericsciple) [inline src/git-auth-helper.ts]: (existing) this calls `git config --local http.https://github.com/.extraheader "AUTHORIZATION: basic ***"`

File: src/git-source-provider.ts [hunk 12]
```diff
@@ -51,7 +43,7 @@ export async function getSource(settings: ISourceSettings): Promise<void> {
 
   // Prepare existing directory, otherwise recreate
   if (isExisting) {
-    await prepareExistingDirectory(
+    await gitDirectoryHelper.prepareExistingDirectory(
```
Comment (ericsciple) [inline src/git-source-provider.ts]: fail when fallback to REST API and SSH, can relax in future if needed

File: __test__/git-auth-helper.test.ts [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+  })
+
+  afterEach(() => {
+    // Unregister mocks
+    jest.restoreAllMocks()
+  })
+
+  afterAll(() => {
+    // Restore RUNNER_TEMP
+    delete process.env['RUNNER_TEMP']
```
Comment (thboop) [inline __test__/git-auth-helper.test.ts]: should we clean up temp after the run as well?
`await io.rmRF(testWorkspace)`

File: src/git-auth-helper.ts [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+    gitCommandManager: IGitCommandManager,
+    gitSourceSettings?: IGitSourceSettings
+  ) {
+    this.git = gitCommandManager
+    this.settings = gitSourceSettings || (({} as unknown) as IGitSourceSettings)
+  }
+
+  async configureAuth(): Promise<void> {
+    // Remove possible previous values
+    await this.removeSsh()
```
Comment (thboop) [inline src/git-auth-helper.ts]: Nit: may want to call removeAuth()

File: src/input-helper.ts [hunk 15]
```diff
@@ -106,6 +106,12 @@ export function getInputs(): ISourceSettings {
   // Auth token
   result.authToken = core.getInput('token')
 
+  // SSH
+  result.sshKey = core.getInput('ssh-key')
+  result.sshKnownHosts = core.getInput('ssh-known-hosts')
+  result.sshStrict =
```
Comment (thboop) [inline src/input-helper.ts]: I think we should default to True if there is no input to match the default value, I also feel safer defaulting to the more secure option

File: __test__/git-auth-helper.test.ts (hunk 13 above)
Comment (ericsciple) [reply __test__/git-auth-helper.test.ts]: i intentionally left the files on disk for debugging. The folder is git-ignored

File: src/git-auth-helper.ts [hunk 16]
```diff
@@ -36,6 +38,8 @@ class GitAuthHelper {
   private readonly tokenPlaceholderConfigValue: string
   private readonly insteadOfKey: string = `url.https://${HOSTNAME}/.insteadOf`
```
Comment (ericsciple) [inline src/git-auth-helper.ts]: dont do insteadOf when sshKey

File: README.md [hunk 17]
```diff
@@ -45,14 +45,40 @@ Refer [here](https://github.com/actions/checkout/blob/v1/README.md) for previous
...
 
-    # Auth token used to fetch the repository. The token is stored in the local git
-    # config, which enables your scripts to run authenticated git commands. The
-    # post-job step removes the token from the git config. [Learn more about creating
-    # and using encrypted secrets](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets)
+    # Personal access token (PAT) used to fetch the repository. The PAT is configured
+    # with the local git config, which enables your scripts to run authenticated git
+    # commands. The post-job step removes the PAT.
+    #
+    # We recommend creating a service account with the least permissions necessary.
```
Comment (thboop) [inline README.md]: NIT: We recommend generating the PAT from a service account with only the necessary permissions. 

may read a little better

File: README.md [hunk 18]
```diff
@@ -45,14 +45,40 @@ Refer [here](https://github.com/actions/checkout/blob/v1/README.md) for previous
...
+    # Also when generating a new PAT, select the least scopes necessary.
+    #
+    # [Learn more about creating and using encrypted secrets](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets)
+    #
     # Default: ${{ github.token }}
     token: ''
 
-    # Whether to persist the token in the git config
+    # SSH key used to fetch the repository. SSH key is configured with the local git
```
Comment (thboop) [inline README.md]: Nit: `The SSH Key is configured...`

File: README.md [hunk 19]
```diff
@@ -45,14 +45,40 @@ Refer [here](https://github.com/actions/checkout/blob/v1/README.md) for previous
...
+    # config, which enables your scripts to run authenticated git commands. The
+    # post-job step removes the SSH key.
+    #
+    # We recommend creating a service account with the least permissions necessary.
```
Comment (thboop) [inline README.md]: Nit: Same as above

File: README.md [hunk 20]
```diff
@@ -73,6 +99,10 @@ Refer [here](https://github.com/actions/checkout/blob/v1/README.md) for previous
 
     # Whether to checkout submodules: `true` to checkout submodules or `recursive` to
     # recursively checkout submodules.
+    #
+    # When the `ssh-key` input is not provided, SSH URLs beginning with
+    # `git@github.com:` are converted to HTTPS.
```
Comment (thboop) [inline README.md]: 🏅

File: __test__/git-auth-helper.test.ts [hunk 21]
```diff
@@ -108,6 +121,52 @@ describe('git-auth-helper tests', () => {
     }
   )
 
+  const configureAuth_copiesUserKnownHosts =
+    'configureAuth copies user known hosts'
+  it(configureAuth_copiesUserKnownHosts, async () => {
+    if (!sshPath) {
```
Comment (thboop) [inline __test__/git-auth-helper.test.ts]: consider a helper function for this

File: __test__/git-auth-helper.test.ts [hunk 22]
```diff
@@ -129,6 +188,173 @@ describe('git-auth-helper tests', () => {
...
+
+    // Assert SSH key
+    const actualSshKeyPath = await getActualSshKeyPath()
+    expect(actualSshKeyPath).toBeTruthy()
+    const actualSshKeyContent = (
+      await fs.promises.readFile(actualSshKeyPath)
+    ).toString()
+    expect(actualSshKeyContent).toBe(settings.sshKey + '\n')
+    if (!isWindows) {
+      expect((await fs.promises.stat(actualSshKeyPath)).mode & 0o777).toBe(
```
Comment (thboop) [inline __test__/git-auth-helper.test.ts]: NIT: Comment what this is

File: __test__/git-auth-helper.test.ts [hunk 23]
```diff
@@ -223,37 +510,135 @@ describe('git-auth-helper tests', () => {
...
+
+      // Assert
+      expect(mockSubmoduleForeach).toBeCalledTimes(1)
+      expect(mockSubmoduleForeach.mock.calls[0][0] as string).toMatch(
+        /unset-all.*insteadOf/
+      )
+    }
+  )
+
+  const configureSubmoduleAuth_doesNotConfigureUrlInsteadOfWhenPersistCredentialsTrueAndSshKeySet =
```
Comment (thboop) [inline __test__/git-auth-helper.test.ts]: Great coverage across all of these different scenarios!

**Comment by ericsciple** [comment]
related to https://github.com/actions/checkout/issues/116

File: README.md (hunk 17 above)
Comment (ericsciple) [reply README.md]: will follow up on next pr

File: README.md (hunk 18 above)
Comment (ericsciple) [reply README.md]: will fix in next pr

File: README.md (hunk 19 above)
Comment (ericsciple) [reply README.md]: will follow up on next pr

File: __test__/git-auth-helper.test.ts (hunk 22 above)
Comment (ericsciple) [reply __test__/git-auth-helper.test.ts]: will follow up on next pr

**Pull request merged**